### PR TITLE
Persist scroll positions to a local file

### PR DESCRIPTION
**Pull request by eshurakov**
### Details
In this PR we're persisting scroll position of notes to a local file in documents folder. To keep the file size small we clean it up by removing deleted notes when the app goes to background.

Ref: #1092

### Test
1. Open any note and scroll, remember where you scrolled.
2. Repeat first step a few times
3. Kill the app and open it again

- [x] Check that notes that you scrolled before open exactly at the same position

### Review
Only one developer is required to review these changes, but anyone can perform the review.

### CODE REVIEW

File: Simplenote/Classes/EditorFactory.swift [hunk 1]
```diff
@@ -44,3 +47,10 @@ class EditorFactory: NSObject {
         return note
     }
 }
+
+
+// MARK: - Constants
+//
+private struct Constants {
+    static let filename = ".scroll-cache"
```
Comment (eshurakov) [inline Simplenote/Classes/EditorFactory.swift:55]: WDYT about filename? 🤔

Comment (jleandroperez) [reply Simplenote/Classes/EditorFactory.swift:55]: I think looks good!!

Comment (jleandroperez) [reply Simplenote/Classes/EditorFactory.swift:55]: (Plus I love the `.` prefix detail!!)

File: Simplenote/Classes/FileStorage.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        return try decoder.decode(T.self, from: data)
+    }
+
+    /// Save an object
+    ///
+    func save(object: T) throws {
+        let data = try encoder.encode(object)
+        try data.write(to: fileURL)
+    }
+}
```
Comment (jleandroperez) [inline Simplenote/Classes/FileStorage.swift:28]: Hey, this is neat!

File: Simplenote/SPAppDelegate+Extensions.swift [hunk 3]
```diff
@@ -355,3 +355,21 @@ private extension SPAppDelegate {
...
+    @objc
+    func cleanupScrollPositionCache() {
+        let allNotes = SPObjectManager.shared().notes()
+        let allIdentifiers: [String] = allNotes.compactMap {
+            guard let note = $0 as? Note, !note.deleted else {
+                return nil
+            }
+
+            return note.simperiumKey
+        }
```
Comment (jleandroperez) [inline Simplenote/SPAppDelegate+Extensions.swift:368]: Suggestion!!

1. Edit `SPObjectManager.h` and set the collection types:
```
- (NSArray<Note *> *)notes;
- (NSArray<Tag *> *)tags;
```

2. Then you can do something like this!
```
let allIdentifiers: [String] = allNotes.compactMap { note in
   note.deleted ? nil : note.simperiumKey
}
```

**Review by jleandroperez — APPROVED** [review]
Solid!! Sent you a tiny suggestion, other than that :shipit: !!

**Review by charliescheer — COMMENTED** [review]
Looks good to me.

File: Simplenote/SPAppDelegate+Extensions.swift (hunk 3 above)
Comment (eshurakov) [reply Simplenote/SPAppDelegate+Extensions.swift:368]: Great suggestion. Love it.

**Comment by eshurakov** [comment]
Thank you!

**Comment by pachlava** [comment]
[testing] fix/feature verified on iPhone XR | iOS 14.4 | Simplenote 4.33.0.0

**Pull request merged**